pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/rule/security/VfHtmlStyleTagXssRule.java

```diff
@@ -0,0 +1,168 @@
+        if (isStyleTag(elementNode) && !isWithinSafeResource) {
+
+            // check if we are within a URL expression
+            if (isWithinUrlMethod(previousText)) {
+                verifyEncodingWithinUrl(node, data);
+            } else {
+                verifyEncodingWithoutUrl(node, data);
+            }
+        }
+    }
+
+    private boolean isStyleTag(ASTElement elementNode) {
+        // are we dealing with HTML <style></style> tag?
+        return STYLE_TAG.equalsIgnoreCase(elementNode.getLocalName());
+    }
+
+    private void verifyEncodingWithinUrl(ASTElExpression elExpressionNode, Object data) {
+
+        // only allow URLENCODING or JSINHTMLENCODING
+        if (escapeDetector.doesElContainAnyUnescapedIdentifiers(
+                elExpressionNode,
+                URLENCODE_JSINHTMLENCODE)) {
+            addViolationWithMessage(
+                    data,
+                    elExpressionNode,
+                    "Dynamic EL content within URL in style tag should be URLENCODED or JSINHTMLENCODED as appropriate");
+        }
+
+    }
+
+    private void verifyEncodingWithoutUrl(ASTElExpression elExpressionNode, Object data) {
+        if (escapeDetector.doesElContainAnyUnescapedIdentifiers(
+                elExpressionNode,
+                ANY_ENCODE)) {
+            addViolationWithMessage(
+                    data,
+                    elExpressionNode,
+                    "Dynamic EL content in style tag should be appropriately encoded");
+        }
+    }
+
+    private boolean isApexPrefixed(ASTElement node) {
+        return node.isHasNamespacePrefix()
+                && APEX_PREFIX.equalsIgnoreCase(node.getNamespacePrefix());
+    }
+
+    /**
+     * Get text content within style tag that leads upto the ElExpression.
+     * For example, in this snippet:
+     * <style>
+     *  div {
+     *   background: url('{!HTMLENCODE(XSSHere)}');
+     * }
+     * </style>
+     *
+     * getPreviousText(...) would return "\n div {\n background: url("
+     *
+     */
+    private String getPreviousText(ASTContent content, ASTElExpression elExpressionNode) {
+        final int indexInParent = elExpressionNode.getIndexInParent();
+        final VfNode previous = indexInParent > 0 ? content.getChild(indexInParent - 1) : null;
+        return previous instanceof ASTText ? previous.getImage() : "";
+    }
+
+    // visible for unit testing
+    static boolean isWithinUrlMethod(String previousText) {
+        // match for a pattern that
+        // 1. contains "url" (case insensitive),
+        // 2. followed by any number of whitespaces,
+        // 3. a starting bracket "("
+        // 4. and anything else but an ending bracket ")"
+        // For example:
+        // Matches: "div { background: url('", "div { background: Url  ( blah"
+        // Does not match: "div { background: url('myUrl')", "div { background: myStyle('"
+
+        return Pattern.compile(URL_METHOD_PATTERN, Pattern.CASE_INSENSITIVE)
```

Comment:
The result of `Pattern.compile` can be the static constant ^^ It's good practice to avoid compiling the pattern every time if possible. Pattern objects are immutable and can be shared (maybe, Apex handles this differently?).

Reply:
@oowekyala I updated this to a static member. 

Roopa is on break, let me know if I can help with this PR.

Thanks.